Code/GraphMol/NewCIPLabelling/Digraph.hpp

```diff
@@ -0,0 +1,258 @@
+      queue.pop();
+
+      if (atom == node->getAtom()) {
+        result.push_back(node);
+      }
+      for (const auto& e : node->getEdges()) {
+        if (!e->isBeg(node)) {
+          continue;
+        }
+        queue.push(e->getEnd());
+      }
+    }
+    return result;
+  }
+
+  /**
+   * Access the reference atom for Rule 6 (if one is set).
+   */
+  A getRule6Ref() const { return rule6Ref; }
+
+  /**
+   * Used exclusively for Rule 6, we set one atom as the reference.
+   * @param ref reference atom
+   */
+  void setRule6Ref(A ref) { rule6Ref = ref; }
+
+  Node<A, B>* init(A atom) {
+    auto visit = std::vector<char>(mol->getNumAtoms());
+    visit[mol->getAtomIdx(atom)] = 1;
+
+    auto atomic_num_num = mol->getAtomicNum(atom);
+    auto atomic_num_den = 1;
+    auto dist = 1;
+    auto flags = 0x0;
+
+    root = new Node<A, B>(this, std::move(visit), atom, atomic_num_num,
+                          atomic_num_den, dist, flags);
+
+    ++numNodes;
+
+    return root;
+  }
+
+  /**
+   * Sets the root node of this digraph by flipping all the
+   * 'up' edges to be 'down'.
+   *
+   * @param newroot the new root
+   */
+  void changeRoot(Node<A, B>* newroot) {
+    auto queue = std::queue<Node<A, B>*>();
+    queue.push(newroot);
+
+    auto toflip = std::vector<Edge<A, B>*>();
+    while (!queue.empty()) {
+      auto node = queue.front();
+      queue.pop();
+
+      for (const auto& e : node->getEdges()) {
+        if (e->isEnd(node)) {
+          toflip.push_back(e);
+          queue.push(e->getBeg());
+        }
+      }
+    }
+    for (auto& e : toflip) {
+      e->flip();
+    }
+    tmproot = newroot;
+  }
+
+  void expand(Node<A, B>* beg) {
+    const A atom = beg->getAtom();
+    const auto& edges = beg->getEdges();
+    const B prev = edges.size() > 0 && !edges[0]->isBeg(beg)
+                       ? edges[0]->getBond()
+                       : nullptr;
+
+    if (MAX_NODE_DIST > 0 && beg->getDistance() > MAX_NODE_DIST) {
+      return;
+    }
+    if (MAX_NODE_COUNT > 0 && numNodes >= MAX_NODE_COUNT) {
+      std::stringstream errmsg;
+      errmsg << "Digraph generation failed: more than " << MAX_NODE_COUNT
+             << "nodes found.";
+      throw TooManyNodesException(errmsg.str());
+    }
+
+    // create 'explicit' nodes
+    for (const B bond : mol->getBonds(atom)) {
+      const A nbr = mol->getOther(bond, atom);
+      const int nbrIdx = mol->getAtomIdx(nbr);
+      const int bord = mol->getBondOrder(bond);
+
+      if (beg->visit[nbrIdx] == 0) {
+        auto end = beg->newChild(nbrIdx, nbr);
+        ++numNodes;
+        addEdge(beg, bond, end);
+
+        // duplicate nodes for bond orders (except for root atoms...)
+        // for example >S=O
+        if (root != beg) {
+          if (mol->getCharge(atom) < 0 &&
+              mol->getFractionalAtomicNum(atom).getDen() > 1) {
+            end =
+                beg->newTerminalChild(nbrIdx, nbr, Node<A, B>::BOND_DUPLICATE);
+            ++numNodes;
```

Comment:
I don't think so. `addEdges()` doesn't really create a new Node, so I don't think it makes sense to put it there.

It might make more sense to put it in the Node constructor, but if we end up using a vector to contain Nodes and Edges, the variable would end up being redundant.

Anyway, besides tests and limiting the size of the graph, it doesn't seem to be used.